### DISCUSSION TITLE
How to identify incoming imports to a virtual module?

### DISCUSSION THREAD

**Discussion by csjh** [rollup/rollup#5287] [Q&A]
I'm trying to make a compiler for a WebAssembly-based language, and I'd like to add treeshaking if possible (before the compilation process into WebAssembly). The problem is, I can't figure out any way to identify the named imports that other files are attempting to import, so that I can forward those on to the compiler and generate the wasm output and JS glue code.

Short version, is there any way to get incoming imports to a virtual module before generating the virtual module code (so, if I understand correctly, before `load()`)?

Relevant issue (I think): #2482

**Comment by lukastaegert** [comment]
That would not be trivial. After all, you cannot know what imports there will be until you parsed EVERYTHING. And what if the module then again imports itself (just a hypothetical idea)? If you mark your import as "external", you might use a prebuilt chunk, but that might not be what you want https://rollupjs.org/plugin-development/#this-emitfile

**Reply by csjh** [reply]
Thanks for the answer! 100% makes sense. My new idea is to use `this.setAssetSource` to defer compilation to `renderChunk` (or other later part of build lifecycle) while just generating the JS glue code in `load`, letting Rollup treeshake that, and analyzing the post-treeshaken output to determine what's still needed.

So now my new problem is figuring out what exports from that file go unused. I was hoping `removedExports` would be my solution but it seems like unused exports don't always end up there. Is there any way to determine those unused exports (short of statically analyzing the output myself)? Or maybe `removedExports` works and there's something wrong in my test:
```js
function myPlugin() {
    const virtualModuleId = "virtual:my-module";
    const resolvedVirtualModuleId = "\0" + virtualModuleId;

    return {
        name: "my-plugin", // required, will show up in warnings and errors
        resolveId(id, importer, options) {
            if (id == virtualModuleId) {
                return resolvedVirtualModuleId;
            }
        },
        load(id) {
            if (id == resolvedVirtualModuleId) {
                const file = this.emitFile({
                    type: "asset",
                    fileName: "virtual.wasm"
                });
                return `
                    const { wasm_export_one, wasm_export_two } = await WebAssembly.instantiateStreaming(fetch(import.meta.ROLLUP_FILE_URL_${file}));


                    // only msg is imported
                    export { wasm_export_one as msg, wasm_export_two as msg2 };
                `;
            }
        },
        generateBundle(options, bundle) {
            // msg2 not in removedExports
            console.log({module_bundle: bundle['assets/index.3e92c44c.js'].modules[resolvedVirtualModuleId]})
        }
    }
}
```

**Reply by csjh** [reply]
Seems like it's possible variable exports aren't included in that? Should be able to work around that by just wrapping the exported functions in another function, which seems to make the unused functions reported in `removedExports`. So now I think I'm just looking for info about how `removedExports` works/what it reports/if it's reliable.

Also, is there any way to find what file a module ends up in that's better than iterating over every key in `bundle` and checking for `file` in its `modules` field?

**Reply by lukastaegert** [reply]
`removedExports` looks at all exports from a file and checks if the variable is included. This is not the same as the export being included as a variable might also be used internally but not in another file. And re-exports are not covered at all.

> Also, is there any way to find what file a module ends up in that's better than iterating

No, unfortunately not.

**Reply by csjh** [reply]
Ahh I see, so in the case of my example, since `wasm_export_two` stays in the bundle, `msg2` isn't considered a removed export? Seems like if I reassign it to its own distinct variable `const msg2 = wasm_export_two;` and export that it's listed in `removedExports` which probably answers my question but just checking.

**Reply by lukastaegert** [reply]
Yes, reassigning to a new variable that is then exported should mark it as a removed export.
